### PR TITLE
Connect worker with twisted endpoints

### PR DESCRIPTION
**Pull request by stelund**
Instead of using an low level TCP connection from a worker to the master use an endpoint. Twisted endpoints offers for example TLS communication and a client can be configured by a connection string instead of using a host and port setting.

Twisted endpoints also offers an reconnect mechanism.

More about endpoints here:
https://twistedmatrix.com/documents/current/core/howto/endpoints.html

Docstrings needs work and documentation.

## Contributor Checklist:

* [x] I have updated the unit tests
* [x] I have created a file in the `master/buildbot/newsfragment` directory (and read the `README.txt` in that directory)
* [x] I have updated the appropriate documentation

### CODE REVIEW

File: worker/buildbot_worker/pb.py [hunk 1]
```diff
@@ -157,22 +143,30 @@ def activity(self, res=None):
...
+
+    maxdelay is deprecated if favour of using twisteds backoffPolicy.
+    """
     Bot = BotPb
 
-    def __init__(self, buildmaster_host, port, name, passwd, basedir,
+    def __init__(self,  buildmaster_host, port, name, passwd, basedir,
                  keepalive, usePTY=None, keepaliveTimeout=None, umask=None,
-                 maxdelay=300, numcpus=None, unicode_encoding=None,
-                 allow_shutdown=None, maxRetries=None):
```
Comment (tardyp) [inline worker/buildbot_worker/pb.py]: why remove maxRetries?

Comment (stelund) [reply worker/buildbot_worker/pb.py]: Ope that must be a merge misstake. I'll fix it later tonight.

**Comment by p12tic** [comment]
Is this change backwards-compatible? I.e. will a worker from, say, buildbot 0.8.9 be able to connect to master after this PR is merged?

**Comment by stelund** [comment]
The master does not have any specific changes to it, it's the worker's point of view that changes.

pbmanager already supports setting an conndescr via the port argument. As can be seen in master/buildbot/pbmanager.py and the test_worker_comm.py.

**Comment by tardyp** [comment]
For me this looks great! @gracinet FYI or FYR

SSL support is a much awaited feature thank you very much for working on it!

**Comment by tardyp** [comment]
Can you add a news fragment, and update the doc with a paragraph on how to setup encryption between master and worker?

We should probably then make encryption the default for new bots

**Comment by gracinet** [comment]
Hey guys. I'm really glad to see this move forward. I'll try and take a look next time I have a computer on hand (should be on the 20th)


Le 15 février 2019 15:22:50 GMT+01:00, Pierre Tardy <notifications@github.com> a écrit :
>Can you add a news fragment, and update the doc with a paragraph on how
>to setup encryption between master and worker?
>
>We should probably then make encryption the default for new bots
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/buildbot/buildbot/pull/4609#issuecomment-464066996

-- 
Envoyé de mon appareil Android avec Courriel K-9 Mail. Veuillez excuser ma brièveté.

**Comment by stelund** [comment]
Yes, I'll write a bit how to setup encryption between master and worker. I want to test it a bit more outside of test-framework.

@gracinet I took the liberty to fixup your commits with pylint and unittest fixes.

**Comment by stelund** [comment]
In worker docs it is written:

> option:: --maxretries
> 
>     This is a number that indicates the maximum number of time the worker will make connection attempts.
>     After that amount, the worker process will stop.
>     This option is useful for :ref:`Latent-Workers` to avoid consuming resources in case of misconfiguration or master failure.
> 
>     For VM based latent workers, the user is responsible for halting the system when Buildbot worker has exited.
>     This feature is heavily OS dependent, and cannot be managed by Buildbot worker.
>     For example with systemd_, one can add ``ExecStopPost=shutdown now`` to the Buildbot worker service unit configuration.

By using backoffPolicy to perform the retry such an option is not included.

https://github.com/twisted/twisted/blob/twisted-16.4.1/twisted/application/internet.py#L487

If the desired instrument by setting maxRetries is to eventually stop the worker and die. Maybe we should keep that. It looks easy enough to implement an backoffPolicy which eventually dies.

.. in buildbot_worker/pb.py
```         endpoint = clientFromString(reactor, conndescr)
+        def policy(attempt):
+            if maxRetries and attempt >= maxRetries:
+                reactor.stop()
+            return backoffPolicy()(attempt)
         pb_service = ClientService(endpoint, bf,
-                                   retryPolicy=backoffPolicy())
+                                   retryPolicy=policy)
         self.addService(pb_service)
```

Another option is to expose backoffPolicy.

**Comment by tardyp** [comment]
`--maxretries` is important indeed

> .. in buildbot_worker/pb.py

I think this implementation is good enough for a first implementation. We can evolve the API later.

**Comment by stelund** [comment]
Right.

What do you think about the argument name conndescr? I saw in the master protocol docs it was referred to as strports from Twisted's module with the same name. But with endpoints twisted seem to forget to have any name for the format.
@tardyp

**Comment by tardyp** [comment]
seriously? they did not reuse strports syntax..

Is this something we could fix? looks like the only difference is /TLS:/ssl/

File: master/docs/relnotes/index.rst [hunk 2]
```diff
@@ -10,6 +10,52 @@ Release Notes
 
 .. towncrier release notes start
 
+Buildbot ``2.0.2-dev64`` ( ``2019-02-18`` )
```
Comment (tardyp) [inline master/docs/relnotes/index.rst]: don't commit index.rst

**Comment by tardyp** [comment]
after looking at the doc, they are indeed the same.

https://twistedmatrix.com/documents/current/api/twisted.application.strports.html

So lets call that endpoint string everywhere, end point to the guide documentation

File: master/docs/manual/installation/worker.rst [hunk 3]
```diff
@@ -176,13 +176,35 @@ Other Worker Configuration
...
+``conndescr``
+    For TLS connections to the master the ``conndescr``-argument must be used to the Worker-initializer. ``buildmaster_host`` and ``port`` must then be ``None``.
+
+    ``conndescr`` should be a string with colon seperated arguments passed to Twisted endpoints ``clientFromString``. An example is ``"TLS:buildbot-master.com:9989"``.
+
+    See more about how to formulate the connection description in `Twisted documentation https://twistedmatrix.com/documents/current/core/howto/endpoints.html`.
+
+    Make sure the worker trusts the masters certificate. If you have an self-signed certificate the trustRoot parameter can be used: ``'TLS:buildbot-master.com:9989:trustRoots=/dir-with-ca-certs'``. It must point to a directory with PEM-encoded certificates in files with file ending .pem.
+
+    ``clientFromString`` supports using certificate and privateKey, but the server does not verify.
```
Comment (tardyp) [inline master/docs/manual/installation/worker.rst]: that is client authentication? please clarify

File: master/buildbot/newsfragments/connect-worker-over-tls.feature [hunk 4]
```diff
@@ -0,0 +1 @@
+:py:class:`~buildbot-worker.buildbot_worker.bot.Worker` now have `conndescr` kw-argument which can be used to connect to an master over TLS.
```
Comment (tardyp) [inline master/buildbot/newsfragments/connect-worker-over-tls.feature]:
```suggestion
:py:class:`~buildbot-worker.buildbot_worker.bot.Worker` now have `connection_string` kw-argument which can be used to connect to an master over TLS.
```

File: master/docs/manual/installation/worker.rst (hunk 3 above)
Comment (stelund) [reply master/docs/manual/installation/worker.rst]: Twisted endpoints clientFromString supports setting client site certificates. Which can be really handy for authentication inte an private network via firewall like Big-IP.

The server side of twisted implementation does not offer to verify who issued the client certificate. Which is lacking for completeness.

Regarding the trustRoots is there no error message if the client does not verify the server. It just tries to reconnect again after 1-2s. Maybe is something that can be fixed.

File: master/docs/manual/installation/worker.rst [hunk 5]
```diff
@@ -176,13 +176,35 @@ Other Worker Configuration
...
     s = Worker(buildmaster_host, port, workername, passwd, basedir,
                keepalive, usepty, umask=umask, maxdelay=maxdelay,
                unicode_encoding='utf-8', allow_shutdown='signal')
 
+.. _Worker-TLS-Config:
+
+Worker TLS Configuration
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+``conndescr``
```
[the remark this answers is not in the document]
Comment (stelund) [reply master/docs/manual/installation/worker.rst]: Good suggestion! I've changed all conndescr too connection_string now.

File: master/docs/manual/configuration/global.rst [hunk 6]
```diff
@@ -422,6 +422,16 @@ This means that you can have the buildmaster listen on a localhost-only port by
...
+*strports* can also be used configure workers connecting over TLS. The syntax is then
+
+.. code-block:: python
+
+   c['protocols'] = {"pb": {"port": "ssl:9989:privateKey=master.key:certKey=master.crt"}}
+
+Please note that IPv6 addresses with : must be escaped with \ as well as : in paths and \ in paths. Read more about the *strports* format in `Twisted documentation https://twistedmatrix.com/documents/current/core/howto/endpoints.html`.
+
+   See also :ref:`Worker TLS Configuration <Worker-TLS-Config>`
+
```
[the remark this answers is not in the document]
Comment (stelund) [reply master/docs/manual/configuration/global.rst]: Thanks for the link!

File: master/docs/manual/installation/worker.rst (hunk 3 above)
Comment (stelund) [reply master/docs/manual/installation/worker.rst]: I've removed the weird bit about using client certificates. It is just confusing.

File: master/docs/manual/installation/worker.rst [hunk 7]
```diff
@@ -183,6 +183,28 @@ Other Worker Configuration
...
+
+    See more about how to formulate the connection string in ConnectionStrings_.
+
+    Make sure the worker trusts the masters certificate. If you have an self-signed certificate the trustRoot parameter can be used: ``'TLS:buildbot-master.com:9989:trustRoots=/dir-with-ca-certs'``. It must point to a directory with PEM-encoded certificates in files with file ending .pem.
+
+.. code-block:: python
+
+    s = Worker(None, None, workername, passwd, basedir,
+               keepalive, connection_string='TLS:buildbot-master.com:9989')
+
```
Comment (tardyp) [inline master/docs/manual/installation/worker.rst]: We should probably  explain that :

```
 s = Worker('master', 9989, workername, passwd, basedir,
                      keepalive)
```

 is equivalent to

``` 
    s = Worker(None, None, workername, passwd, basedir,
                      keepalive, connection_string="tcp:master:9989")
```

File: master/docs/relnotes/index.rst [hunk 8]
```diff
@@ -10,6 +10,48 @@ Release Notes
 
 .. towncrier release notes start
 
+Buildbot ``2.0.2-dev66`` ( ``2019-02-19`` )
```
Comment (tardyp) [inline master/docs/relnotes/index.rst]: please remove this chunk from the commit. This is autogenerated code.

Comment (stelund) [reply master/docs/relnotes/index.rst]: removed

File: master/docs/manual/installation/worker.rst (hunk 7 above)
Comment (stelund) [reply master/docs/manual/installation/worker.rst]: I've added that too. And also expanded about the trustRoots.

**Comment by stelund** [comment]
It shows up as 1 review request changes.. but I cant see it. Do you @tardyp ?

**Comment by tardyp** [comment]
I think this is good to go for now for revieweability sake. A few nitpicks that could be done in a later PR:

- update the buildbot.tac generation to not use deprecated way
- deprecate the master, port first arguments (not sure how)

And other more general usuability of encryption, generation of certificate is not very user friendly, and we want to facilitate encryption.
- self signed certificate autogeneration at buildbot create
- auto trust of self signed certificate at buildbot-worker create


Anyway, thank you very much for this cool work!

**Pull request merged**